feat: ✨ post on why to use Rust for making software (#279)
seedcase-project/decisions: why-rust-for-building-software/index.qmd

---
title: "Why Rust for building software"
description: |
  We want to build software, specifically for command-line and web applications,
  in a language that is sustainable (both in our time but also in resources),
  that is secure and reliable, and that has good developer ergonomics. This post
  describes why we decided to switch from Python to Rust for building software.
date: "2026-06-01"
categories:
  - develop
---

## Context and problem statement

::: content-hidden
State the context and some background on the issue, then write a statement in
the form of a question for the problem.
:::

Building and designing stable, secure, and reliable software is hard. Really
hard. And the needs for software built now are different and more complex than

> they were in the past, and even just a few years ago. Especially with the rise

Reviewer: ```suggestion
they were in the past, even just a few years ago. Especially with the rise
```

of AI coding assistants and with it an increase in the amount of software being

> built that is of varying quality and importantly security. Supply chain attacks,

Reviewer: ```suggestion
of AI coding assistants, which has led to an increase in the amount of low-quality software with subpar security practices. Supply chain attacks,
```

> built that is of varying quality and importantly security. Supply chain attacks,

Author: ```suggestion
of AI coding assistants, which has led to an increase in the amount of lower-quality software with inadequate security practices. Supply chain attacks,
```

less focus on software security and more on volume, and vulnerabilities and bugs

> are becoming more common. Combined with the fact that AI coding assistants can

Reviewer: ```suggestion
prioritizing volume over security, vulnerabilities, and bugs
are all becoming more common. Combined with the fact that AI coding assistants can
```

easily generate code, the importance of strictness and correctness in the
program built and in having strong guardrails to enforce that the software is

> well-built are increasingly important.

Reviewer: ```suggestion
easily generate code, the importance of having strong guardrails to enforce software strictness and correctness is increasingly important.
```

On top of that, we've refined our
[values](https://design.seedcase-project.org/software/) for building software

> that includes explicitness, a consideration for sustainability (meaning using as

Reviewer: ```suggestion
to include explicitness, a consideration for sustainability (meaning using as
```

little computing resources as possible), and valuing the code itself just as
much as the product built with it. We also have more ideas for different
software tooling we could build, most of which is mainly command-line based.

> This means we have a clearer idea for what we need and want. We originally

Reviewer: ```suggestion
much as the product built with it. As we have been working on the Seedcase ecosystem,
we have also refined our needs and wants regarding which software we are going to build.
We originally
```

decided on [Python for building software and processing
data](/why-python/index.qmd), however as we gained experience and learned more
about what we needed and will need for future products, we realised Python no
longer has a good fit for most of our needs and our aims. This post updates our

> decision with regard to these changes and our current needs.

Reviewer: ```suggestion
decided on [using Python for building software and processing
data](/why-python/index.qmd), but it is no longer
the best fit for our updated needs and values. This post updates our
decision in regards to these changes.
```

  understand.
- Has support and an ecosystem for building command-line applications as well as
  (simple) web applications (e.g. for client-side/in-browser processing).
- Better investment to learn and build skills in for a wider variety of products
  that we may build in the future.

## Considered options

::: content-hidden
List and describe some of the options, as well as some of the benefits and
drawbacks for each option.
:::

Since Python no longer fits our needs (e.g. doesn't have built-in and enforced
types, the developer ergonomics is fragmented, and isn't as easy to build web
applications with), we didn't include it in our consideration. And while we
previously considered R, Java, and C++ in our [previous
decision](/why-python/index.qmd), we didn't include them in this post because
they fit our needs even less than Python does (see the brief notes below on
why).

- [Rust](https://www.rust-lang.org/)

We didn't consider these languages because they don't fit our needs as well as
the above options, nor do they have a good match with our values (e.g. not a
common language in developer/researcher tooling nor as common in
academic/research settings):

> - [Go](https://golang.org/): Designed primarily for building web applications.

Reviewer: I actually think Go would be a strong candidate for the CLI framing we gave above. My understanding it that Go's CLI framework Cobra is one of the most capable ones out there and it is the backbone for e.g. the github CLI and Kubernetes. If anything it seems like its ability to deal with some web logic  would be a weakness. (I'm not saying we should use Go, just rephrase this).

Author: From what I understand, Go has very good functionality for web-based stuff, including CLIs that connect with/interact with web resources/APIs. So it isn't that it is good at CLIs, it's good at CLIs that connect with the internet, like `gh cli`.

```suggestion
- [Go](https://golang.org/): Designed primarily for building applications that build on or interact with the web.
```

  and distribute software, which makes the developer experience very nice and
  ergonomic. There is no need to decide on a specific tool for a specific task,
  as it is all built-in.
- While not as common as many other languages, because it is so low level (but
  with high-level ergonomics in the language design), it can create bindings for
  many other languages, including to Python and to R.
- Recently, a lot of developer tooling in many languages including JavaScript
  (e.g. [Biome](https://biomejs.dev)), Python (e.g.
  [uv](https://docs.astral.sh/uv) or [Ruff](https://docs.astral.sh/ruff)), and R
  (e.g. [Air](https://posit-dev.github.io/air/) and
  [Jarl](https://jarl.etiennebacher.com)), have been built in Rust.
- Rust is becoming more common in academic and research settings, and there are
  more and more libraries being built in Rust e.g. for data processing with
  [Polars](https://www.pola.rs/). And because of the ability to create bindings
  for other languages, some research libraries are being built in Rust to allow
  for cross-language use.
- The compiler and checker provides very good messages that teach you about the
  issue as well as showing how to fix it.

> - Because of it's ownership model, as well as very strict and modern design, it

Reviewer 2: ```suggestion
- Because of its ownership model, as well as very strict and modern design, it
```

  storage resources, which is better for sustainability especially for software
  that is designed to be run often.
- Improves developer experience because of the availability of built-in tools
  and to strictness in the compiler checks, which means as a team you don't have
  to make as many decisions about tooling and how the code should be written.
- With more AI coding assistants, having a language that enforces good practices
  and strong guardrails means that the code generated by AI assistants is more
  likely to be better quality compared to a language that doesn't have those
  guardrails, which means less time spent fixing and debugging code generated by
  AI assistants.
- Because of the strict compiler checks, refactoring code is much easier and
  safer as it ensures that any changes made don't break code in unexpected ways.
- The distribution system with `cargo` stops certain types of supply chain
  attacks commonly seen in e.g. npm and PyPI. For example, it locks dependencies
  to specific versions of packages and once a package is published, it can not
  be removed (it is immutable).
:::::
::::: {.column}
#### Drawbacks

> - Because of it's strictness and lower-level nature, it has a steeper learning

Reviewer 2: ```suggestion
- Because of its strictness and lower-level nature, it has a steeper learning
```

  curve compared to other languages.
- It still is a newer language, and while it is growing in popularity, it
  doesn't have as large of a community or ecosystem as some other languages,
  which means there may be less libraries available for specific tasks.
- Analysing Rust code during compilation checks can be slower than other
  languages and can take up a lot of RAM when, e.g. working in an IDE with
  [`rust-analyzer`](https://rust-analyzer.github.io/), which can be a problem
  for larger codebases and for developers with less powerful computers.
- Running checks on continuous integration workflows can be slower because of
  the analyzer.
- Rust isn't as common in academic and research settings, which means there will

>   be less potential for external contributions from these communities.

Reviewer: Should we include anything in the pros and concs about the benefits and costs of learning something new and familiar for the team?

Maybe there is also a drawback related to using multiple languages between libraries within the same project? I'm thinking of e.g. if propagate (or another rust package) might need a function from soil (or similar). Or if there isn't, maybe a note about that we considered this and don't think it is a concern (as per your comment in https://github.com/seedcase-project/seedcase-propagate/issues/6#issuecomment-4591985450) to show that we took it into account.

Author: I wouldn't consider this a drawback of Rust specifically, which is what this section is about. That might go into our consequences section.

Reviewer: That makes sense 👍  I was thinking of this section as drawbacks for us adopting Rust rather than drawbacks of Rust in general (that are relevant to us). I will make a PR to include this in the template to avoid future confusion. Edit: Done in https://github.com/seedcase-project/decisions/pull/283

:::::
:::

## Decision outcome

::: content-hidden
What decision was made, use the form "We decided on CHOICE because of REASONS."
:::

There weren't that many options to choose from other than Rust. If Rust wasn't
available, we'd have to manage with Python and its drawbacks until a language
that fit our needs eventually came along. However, Rust does exist and fits a
lot of our needs and values, as well has providing a lot of potential for taking
on more complex and impactful projects in the future. Therefore, we decided to
switch to Rust specifically for our command-line and web applications, while
still using Python for data processing tasks and for building libraries that are
meant to provide functionality to simplify those data processing tasks.

> ### Consequences

Reviewer: What's the difference between "consequences" and the "benfits/drawbacks" above? It seems like all of these could slot in under the headings above instead? The last point even occurs in both sections.

Author: Consequences is about what happens to us because of this decision while drawback is something that is unique to the language itself/the community around it. E.g. the last points that you mention, in the drawbacks its *generally* less contributors, and for consequences its *we likely won't get contributors*.

Reviewer: Hmmm, I see... With my current understanding of the intended distinction, it seems like most items from "Drawbacks" (and "Benefits") could then just be repeated here with some rephrasing, which might not be that useful?

I wonder if we can emphasize this distinction by commenting more elaborately on how a general drawback might (or might not) affect us. We could also only include those items where there is valuable additional information that is specific to us and drop those that are more just a rephrase of the general drawback? I made an attempt below and an example.

::: content-hidden
List some potential consequences of this decision.
:::

- Because of the smaller ecosystem, we may have to build more of our own tools
  and libraries.
- Rust is known to have a steep learning curve, which means we need to invest
  time in learning it and will slow down our development speed. However, like
  learning any new skill, once learned, development will speed up.
- We might become annoyed with the slower compilation times and higher local
  resource usage during development. But this is a tradeoff for the benefits for
  strictness and safety. Plus, there are many tools and techniques to mitigate
  this.
- Rust isn't as common in research settings so we will not get as many
  contributions and external collaborators. However, this might not be an issue
  for us. For one, there's a strong trend for more Rust tools and libraries
  being used in research settings because of its benefits, so this may change in

>   the future. For another, getting contributions is difficult for any language,

Reviewer 3: ```suggestion
  the future. Getting contributions is difficult for any language,
```

  and even more so in research settings, so even from the start we never

>   expected many external contributions.

Reviewer: Maybe it would be interesting to mention somewhere if we are planning to translate any existing tools to rust or if they all fall under the "data" too category? E.g. flower clearly is CLI-heavy but also data-related?

Author: Hm, I don't feel that fits with a decision post, that's more a roadmap/planning thing. The decision post should ideally not relate to what we plan to do, aside from using it for the purposes listed here.

---
title: "Why Rust for building software"
description: |
  We want to build software, specifically for command-line and web applications,

>   in a language that is sustainable (both in our time but also in resources),

Reviewer 3: What is meant by "resources" here?

>   that is secure and reliable, and that has good developer ergonomics. This post

Reviewer 3: ```suggestion
  secure and reliable, and that has good developer ergonomics. This post
```

>   describes why we decided to switch from Python to Rust for building software.

Reviewer 3: ```suggestion
  describes why we decided to switch from building software in Python to building in Rust.
```
Maybe?

date: "2026-06-01"
categories:
  - develop
---

## Context and problem statement

::: content-hidden
State the context and some background on the issue, then write a statement in
the form of a question for the problem.
:::

Building and designing stable, secure, and reliable software is hard. Really
hard. And the needs for software built now are different and more complex than
they were in the past, even just a few years ago. Especially with the rise of AI
coding assistants, which has led to an increase in the amount of lower-quality

> software with inadequate security practices. Supply chain attacks, prioritizing

Reviewer 3: ```suggestion
they were in the past, even just a few years ago. Especially the rise of AI
coding assistants has led to an increase in the amount of lower-quality
software with inadequate security practices. Supply chain attacks, prioritizing
```

> volume over security, vulnerabilities, and bugs are all becoming more common.

Reviewer 3: Have bugs become more common or were they always common? 🤔

Author: Hard thing to say... Has the percentage been the same but now there's more code/software out there? Without doing extensive systematic analysis, it would be nearly impossible to determine. From anecdotal experiences, it seems like there are more.

> Combined with the fact that AI coding assistants can easily generate code, the

Reviewer 3: ```suggestion
Combined with the fact that AI coding assistants can generate code faster than people can understand it, the
```
Maybe?

importance of having strong guardrails to enforce software strictness and
correctness is increasingly important.

On top of that, we've refined our
[values](https://design.seedcase-project.org/software/) for building software to
include explicitness, a consideration for sustainability (meaning using as

> little computing resources as possible), and valuing the code itself just as

Reviewer 3: 🌱 🌍

much as the product built with it. As we have been working on the Seedcase

> ecosystem, we have also refined our needs and wants regarding which software we

Reviewer 3: ```suggestion
ecosystem, we have also refined our needs and wants regarding the software we
```

are going to build. We originally decided on [using Python for building software
and processing data](/why-python/index.qmd), but it is no longer the best fit
for our updated needs and values. This post updates our decision in regards to
these changes.

## Decision drivers

::: content-hidden
List some reasons for why we need to make this decision and what things have
arisen that impact work.
:::

> Many of the drivers are the same as in our previous decision, except we have

Reviewer 3: Add the original ones here as well, so this can stand alone?

> additional ones:

Reviewer 3: ```suggestion
the following additional ones. We want a language that provides:
```

- Better developer ergonomics and built-in development tooling, such as for
  building, testing, and distributing software.

> - More strict checks on the code, to prevent easily solvable bugs and security

Reviewer 3: ```suggestion
- Great developer ergonomics and built-in development tooling, such as for
  building, testing, and distributing software.
- Strict checks on the code, to prevent easily solvable bugs and security
```

  issues from the beginning.

> - Built-in and required types, to make the code more explicit and easier to

Reviewer 3: ```suggestion
- Built-in and required types to make the code more explicit and easier to
```

  understand.

> - Has support and an ecosystem for building command-line applications as well as

Reviewer 3: ```suggestion
- Support and an ecosystem for building command-line applications as well as
```

  (simple) web applications (e.g. for client-side/in-browser processing).
- Better investment to learn and build skills in for a wider variety of products

>   that we may build in the future.

Reviewer 3: I understand the point, but not necessarily this sentence. Maybe a slight rewrite could help?
```suggestion
- Skill development and learning that can be reused across a broad range of future products.
```

## Considered options

::: content-hidden
List and describe some of the options, as well as some of the benefits and
drawbacks for each option.
:::

Since Python no longer fits our needs (e.g. doesn't have built-in and enforced
types, the developer ergonomics is fragmented, and isn't as easy to build web

> applications with), we didn't include it in our consideration. And while we

Reviewer 3: To me, it would make sense to keep Python as a consideration bc it used to fit our needs. Adding it here would offer explicit pros/cons for why that isn't try anymore and why Rust is a better choice for us now. Currently, the pros and cons are kind of spread across this document "hidden" within the different sections; if it was a considered option, it would be easier to compare directly with Rust.

previously considered R, Java, and C++ in our [previous
decision](/why-python/index.qmd), we didn't include them in this post because
they fit our needs even less than Python does (see the brief notes below on
why).

> - [Rust](https://www.rust-lang.org/)

Reviewer 3: It does feel a little artificial only considering one option 🤔

We didn't consider these languages because they don't fit our needs as well as

> the above options, nor do they have a good match with our values (e.g. not a

Reviewer 3: ```suggestion
the above option, nor do they have a good match with our values (e.g. not a
```
Currently only one option :grin:

  isn't suited for building more complex and diverse applications.

### Rust

Rust is a relatively newer language, having been [first publicly
released](https://en.wikipedia.org/wiki/Rust_(programming_language)) in 2012. It
is a "systems programming language", meaning it is designed to be low-level and
close to the hardware. But it also has high-level ergonomics and safety features
typically seen in higher-level languages. It is a statically typed language,
which means that types are checked at compile time, mostly implements functional
programming principles, and has a unique ownership model that ensures memory
safety without needing a garbage collector.

::: {.columns}
::::: {.column}
#### Benefits

- It's consistently ranked as one of the [most loved programming
  languages](https://survey.stackoverflow.co/2023/#section-admired-and-desired-programming-scripting-and-markup-languages)
- Strong and enforced type system, which removes whole classes of bugs and
  security issues.
- Memory ownership system, which entirely removes memory leaks and
  safety/security issues related to that.
- The built-in `cargo` tool contains basically everything needed to build, test,

>   and distribute software, which makes the developer experience very nice and

Reviewer 3: Maybe align the list item phrasing here. E.g., the first one starts with "It's" which makes me want to have a subejct in the following items as well, "It has a strong and enforced ..." , "It includes a memory ownership system ...", etc.

  ergonomic. There is no need to decide on a specific tool for a specific task,
  as it is all built-in.
- While not as common as many other languages, because it is so low level (but
  with high-level ergonomics in the language design), it can create bindings for

>   many other languages, including to Python and to R.

Reviewer 3: ```suggestion
- It can create bindings for
  many other languages, including to Python and to R.
```
Maybe?

- Recently, a lot of developer tooling in many languages including JavaScript
  (e.g. [Biome](https://biomejs.dev)), Python (e.g.
  [uv](https://docs.astral.sh/uv) or [Ruff](https://docs.astral.sh/ruff)), and R
  (e.g. [Air](https://posit-dev.github.io/air/) and

>   [Jarl](https://jarl.etiennebacher.com)), have been built in Rust.

Reviewer 3: ```suggestion
- Recently, a lot of developer tooling in many languages have been built in Rust, including JavaScript
  (e.g. [Biome](https://biomejs.dev)), Python (e.g.
  [uv](https://docs.astral.sh/uv) or [Ruff](https://docs.astral.sh/ruff)), and R
  (e.g. [Air](https://posit-dev.github.io/air/) and
  [Jarl](https://jarl.etiennebacher.com)).
```

- Rust is becoming more common in academic and research settings, and there are
  more and more libraries being built in Rust e.g. for data processing with

>   [Polars](https://www.pola.rs/). And because of the ability to create bindings

Reviewer 3: ```suggestion
  [Polars](https://www.pola.rs/). Because of the ability to create bindings
```

  for other languages, some research libraries are being built in Rust to allow
  for cross-language use.
- The compiler and checker provides very good messages that teach you about the
  issue as well as showing how to fix it.
- Because of its ownership model, as well as very strict and modern design, it
  requires learning and thinking more about how code is built, how computers
  work, and how to write good and safe code, which builds developer experience,
  skill, and knowledge.
- Because it builds to native binaries and doesn't need a runtime interpreter,
  it is very fast and efficient. This means that it requires less computing and

>   storage resources, which is better for sustainability especially for software

Reviewer 3: ```suggestion
  storage resources, which is more sustainable, especially for software
```

  that is designed to be run often.
- Improves developer experience because of the availability of built-in tools
  and to strictness in the compiler checks, which means as a team you don't have
  to make as many decisions about tooling and how the code should be written.
- With more AI coding assistants, having a language that enforces good practices
  and strong guardrails means that the code generated by AI assistants is more
  likely to be better quality compared to a language that doesn't have those
  guardrails, which means less time spent fixing and debugging code generated by
  AI assistants.
- Because of the strict compiler checks, refactoring code is much easier and
  safer as it ensures that any changes made don't break code in unexpected ways.
- The distribution system with `cargo` stops certain types of supply chain
  attacks commonly seen in e.g. npm and PyPI. For example, it locks dependencies
  to specific versions of packages and once a package is published, it can not
  be removed (it is immutable).

> :::::

Reviewer 3: Wait what, you can't remove dependencies even in newer versions of packages?

Author: Yea, all versions of packages published to crate.io are fixed and can't be removed, unlike PyPI or npm or CRAN. The dependencies are locked when published at a given version. If you update the dependencies and than create a new version, those new dependency versions are included.

::::: {.column}
#### Drawbacks

- Because of its strictness and lower-level nature, it has a steeper learning
  curve compared to other languages.
- It still is a newer language, and while it is growing in popularity, it
  doesn't have as large of a community or ecosystem as some other languages,

>   which means there may be less libraries available for specific tasks.

Reviewer 3: Oh, interesting to see whether we'll run into some gaps here.

- Analysing Rust code during compilation checks can be slower than other
  languages and can take up a lot of RAM when, e.g. working in an IDE with
  [`rust-analyzer`](https://rust-analyzer.github.io/), which can be a problem
  for larger codebases and for developers with less powerful computers.
- Running checks on continuous integration workflows can be slower because of

>   the analyzer.

Reviewer 3: CRAN check slow? 😝 👀

- Rust isn't as common in academic and research settings, which means there will
  be less potential for external contributions from these communities.
:::::
:::

## Decision outcome

::: content-hidden
What decision was made, use the form "We decided on CHOICE because of REASONS."
:::

There weren't that many options to choose from other than Rust. If Rust wasn't
available, we'd have to manage with Python and its drawbacks until a language
that fit our needs eventually came along. However, Rust does exist and fits a
lot of our needs and values, as well has providing a lot of potential for taking
on more complex and impactful projects in the future. Therefore, we decided to
switch to Rust specifically for our command-line and web applications, while
still using Python for data processing tasks and for building libraries that are
meant to provide functionality to simplify those data processing tasks.

### Consequences

::: content-hidden
List some potential consequences of this decision.
:::

- Because of the smaller ecosystem, we may have to build more of our own tools

>   and libraries. While this could become laborious, we believe that it will be a valuable asset for individual team members to learn and develop their value as software engineers.

Reviewer 3: Formatting is needed here 🎨

  Selecting the tools that help us build the best possible software for
  a given set of tasks is more likely to attract contributors in the long run.
  There's also a recent strong trend for more research-based tools and
  libraries built in Rust,
  so we expect the potential contributor and collaborator pool
  to grow.
- It might become difficult for us to build most software in Rust, while writing
  code to process data in Python, as this will require context switching
  between languages. However, this is also a chance for us to practice
  working in multiple languages and to learn that we're capable of using the right
  tool for the right task, even if it isn't something we're familiar with or
  have expertise in. Widening our expertise and learning about best practices
  in Rust could also bring about positive benefits for our Python coding habits.
- There may be some functionality that we've already developed in Python that we
  would like to also use in a Rust-based tool. This shouldn't be a major issue
  as we've built most of our software as command-line tools, rather than as libraries with reusable
  functionality. If library code needs to be shared and we don't want to

>   re-implement it in rust, we can use [PyO3](https://github.com/PyO3/pyo3) to

Reviewer 3: ```suggestion
  re-implement it in Rust, we can use [PyO3](https://github.com/PyO3/pyo3) to
```

- Running checks on continuous integration workflows can be slower because of
  the analyzer.
- Rust isn't as common in academic and research settings, which means there will
  be less potential for external contributions from these communities.
:::::
:::

## Decision outcome

::: content-hidden
What decision was made, use the form "We decided on CHOICE because of REASONS."
:::

There weren't that many options to choose from other than Rust. If Rust wasn't
available, we'd have to manage with Python and its drawbacks until a language
that fit our needs eventually came along. However, Rust does exist and fits a
lot of our needs and values, as well has providing a lot of potential for taking
on more complex and impactful projects in the future. Therefore, we decided to
switch to Rust specifically for our command-line and web applications, while
still using Python for data processing tasks and for building libraries that are
meant to provide functionality to simplify those data processing tasks.

### Consequences

::: content-hidden
List some potential consequences of this decision.
:::

- Because of the smaller ecosystem, we may have to build more of our own tools

>   and libraries.

Reviewer: ```suggestion
- Because of the smaller ecosystem, we may have to build more of our own tools
  and libraries. While this could become laborious, we believe that it will be a valuable asset for individual team members to learn and develop their value as software engineers.
```

- Rust is known to have a steep learning curve, which means we need to invest
  time in learning it and will slow down our development speed. However, like
  learning any new skill, once learned, development will speed up.
- We might become annoyed with the slower compilation times and higher local
  resource usage during development. But this is a tradeoff for the benefits for
  strictness and safety. Plus, there are many tools and techniques to mitigate

>   this.

Reviewer: ```suggestion
```

These two are basically a rehash of what is written in the general drawback section without any additional comments on how they would apply to us specifically, so I think they are fine to drop.

- Because Rust isn't as common in research settings, we probably won't get as
  many contributions and external collaborators. However, this might not be an
  issue for us. For one, there's a strong trend for more Rust tools and
  libraries being used in research settings because of its benefits, so this may
  change in the future. For another, getting contributions is difficult for any
  language, and even more so in research settings, so even from the start we

>   never expected many external contributions.

Reviewer: ```suggestion
- Getting potentially fewer external contributions and collaborators
  is not a major concern since we generally expect this number to be
  close to zero for a new tool written in any language,
  particularly in a research setting like this.
  Selecting the tools that help us build the best possible software for
  a given set of tasks is more likely to attract contributors in the long run.
  There's also a recent strong trend for more research-based tools and
  libraries built in Rust,
  so we expect the potential contributor and collaborator pool
  to grow.
```

This is a good example of what I'm referring to in my post above in terms of adding additional context about how a potential drawback applies or doesn't apply to us specifically.

- It might be harder for us to build most of our software in Rust, but write
  code to process data in Python, which will require some context switching
  between languages. However, this is also good practice for us to learn how to
  work in multiple languages and to learn that we're capable of using the right
  tool for the right task, even if it isn't something we're familiar with or

>   expertise in.

Reviewer: ```suggestion
- It might become difficult for us to build most software in Rust, while writing
  code to process data in Python, as this will require context switching
  between languages. However, this is also a chance for us to practice
  working in multiple languages and to learn that we're capable of using the right
  tool for the right task, even if it isn't something we're familiar with or
  have expertise in. Widening our expertise and learning about best practices 
  in Rust could also bring about positive benefits for our Python coding habits.
```

- There may be some functionality that we've developed in Python that we won't
  be able to use in Rust. This isn't a major issue, as most of our software so
  far has been built as command-line tools, rather than libraries with reusable

>   functionality.

Reviewer: ```suggestion
- There may be some functionality that we've already developed in Python that we 
  would like to also use in a Rust-based tool. This shouldn't be a major issue 
  as we've built most of our software as command-line tools, rather than as libraries with reusable
  functionality. If library code needs to be shared and we don't want to
  re-implement it in rust, we can use [PyO3](https://github.com/PyO3/pyo3) to
  call one language from the other.
```

- [JavaScript](https://developer.mozilla.org/en-US/docs/Web/JavaScript) or
  [TypeScript](https://www.typescriptlang.org/): These are mainly used for web
  development. They have a much more fragmented tooling ecosystem. Of the other
  languages, they experience the most security issues and bugs (mostly with
  JavaScript), in particular having had the most supply chain attacks against
  the ecosystem. Both are not known for being secure nor fast languages.
- [R](https://www.r-project.org/): Mainly used for data analysis and statistics.
  It isn't designed nor suited for building more complex and diverse
  applications.

### Python

[Python](https://www.python.org/) is a high-level, interpreted, general-purpose
programming language. It is known for its readability and simplicity.
Originally, it was designed to be used as a scripting language and "glue"
between other tools and languages, but since then has evolved to be used for
doing a wider variety of tasks, including building software.

::: callout-note
Many of these items were taken from the [previous
decision](/why-python/index.qmd) post. More drawbacks have been added that we've

> experienced.

Reviewer 3: ```suggestion
decision](/why-python/index.qmd) post. Below, we've added more drawbacks that we've
experienced when building software (libraries and CLIs) with Python.
```

:::

::: columns
::::: column
#### Benefits

- It is a concise, easy-to-read, beginner-friendly syntax. It prioritizes
  readability and simplicity over many other aspects of the language.
- Iterating and prototyping can be done much more quickly because of its
  interpreted nature, simple syntax, and dynamic typing (types are
  checked/assigned at runtime) compared to other languages.

> - There is a very active and massive community, with a wide range of extensive

Reviewer 3: ```suggestion
- It has is a very active and massive community with a wide range of extensive
```

  difficult to find and catch certain types of bugs and issues. Type-hints can
  help but they are only labels and don't actually do anything aside from
  annotating the code. This means some bugs are very difficult to identify until
  it actually runs.
- Dynamic type-checking at runtime can make it more difficult to catch certain
  types of bugs early on.
- Python's roots as a scripting language means that some of its core designs are
  not designed for effectively and easily building secure and reliable software.
  For example, there are several ways to build a package ("flat layout" vs
  "`src/` layout") and modules are not truly isolated (e.g. circular imports are
  possible and modules can modify the global state of other modules).
- The developer ergonomics and ecosystem is very fragmented. The Python
  Authority doesn't handle or develop specific tools for building, testing, and
  distributing software. This ends up leading to different groups building their
  own tools. For example, there are more than a dozen different code linter and
  formatters or up to a dozen different packaging and distribution tools. That

>   leads to "analysis paralysis" for just determining which tool to use for any

Reviewer 3: ```suggestion
  can lead to "analysis paralysis" just for determining which tool to use for any
```

  [uv](https://docs.astral.sh/uv) or [Ruff](https://docs.astral.sh/ruff)), and R
  (e.g. [Air](https://posit-dev.github.io/air/) and
  [Jarl](https://jarl.etiennebacher.com)).
- Rust is becoming more widely used in academic and research settings, and there
  are more and more libraries being built in Rust e.g. for data processing with
  [Polars](https://www.pola.rs/). Because of the ability to create bindings for
  other languages, some research libraries are being built in Rust to allow for
  cross-language use.
- The compiler and checker provides detailed and informative messages that teach
  you about the issue as well as showing how to fix it.
- Because of its ownership model and strict and modern design requires thinking
  more deeply about how code is built, how computers work, and how to write good
  and safe code. This builds and deepens developer experience, skill, and
  knowledge.
- Because it builds to native binaries and doesn't need a runtime interpreter,
  it is very fast and efficient. This means that it requires less computing and
  storage resources, which is more sustainable, especially for software that is
  designed to be run often.

> - There is a better developer experience because of the availability of built-in

Reviewer 3: ```suggestion
- The developer experience is better because of the availability of built-in
```
